File: f1/f1-socket.py
import json
import socket

# set reusable address
s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
s.bind(('0.0.0.0', 8080))
s.listen(5)
print("Server listening on port 8080")

client_bytes = b""
client_socket, client_address = s.accept()
while True:
    print(f"Connection from {client_address}")
    recv_data = client_socket.recv(1024)
    if b"\r\n\r\n" in recv_data:
        client_bytes += recv_data[:len(recv_data)-len(b"\r\n\r\n")]
        break
    client_bytes += recv_data

response_body = json.dumps({"data": client_bytes.decode(), "status": "ok"})

response = f"HTTP/1.1 200 OK\r\nContent-Length: {len(response_body.encode())}\r\nContent-Type: application/json\r\n\r\n{response_body}".encode()
client_socket.sendall(response)
client_socket.close()
s.close()

File: gcp/f1/f1-socket.go
package main

import (
	"bytes"
	"encoding/json"
	"fmt"
	"net"
)

type Response struct {
	Data   string `json:"data"`
	Status string `json:"status"`
}

func main() {
	delimiter := []byte("\r\n\r\n")
	ln, err := net.Listen("tcp", ":8080")
	if err != nil {
		fmt.Println("Error starting server:", err)
		return
	}
	defer ln.Close() // Ensure the listener is closed when the main function exits
	fmt.Println("Server listening on port 8080")

	clientConn, err := ln.Accept()
	if err != nil {
		fmt.Println("Error accepting connection:", err)
		return
	}
	defer clientConn.Close() // Ensure the client connection is closed when done
	fmt.Println("Accepted connection from", clientConn.RemoteAddr())
	buffer := make([]byte, 4096)
	clientData := make([]byte, 0, 4096)
	for {
		n, err := clientConn.Read(buffer)
		fmt.Println("Received data of", n, "bytes: ", string(buffer[:n]))
		if err != nil {
			fmt.Println("Error reading from connection:", err)
			break
		}
		if n == 0 {
			fmt.Println("Connection closed by client")
			break
		}
		clientData = append(clientData, buffer[:n]...)
		if bytes.Contains(clientData, delimiter) == true {
			fmt.Println("Received complete message:", string(clientData))
			fmt.Println("Total bytes:", len(clientData))
			break
		}
	}

	response := Response{
		Data:   string(clientData),
		Status: "ok",
	}
	fmt.Println("Response prepared:", response)
	responseVal, err := json.Marshal(response)
	if err != nil {
		fmt.Println("Error marshaling response:", err)
		return
	}

	// build the response
	fullResponse := fmt.Sprintf("HTTP/1.1 200 OK\r\nContent-Length: %v\r\nContent-Type: application/json\r\n\r\n%v", len(responseVal), string(responseVal))

	fmt.Println("Response JSON:", fullResponse)
	_, err = clientConn.Write([]byte(fullResponse))
	if err != nil {
		fmt.Println("Error writing response to client:", err)
		return
	}

}
